x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/TrainedModelProvider.java

```diff
@@ -213,20 +223,21 @@ private TrainedModelDefinition parseModelDefinitionDocLenientlyFromSource(BytesR
         }
     }
 
-    private void indexObject(String docId, ToXContentObject body, ActionListener<IndexResponse> indexListener) {
+    private IndexRequest createRequest(String docId, ToXContentObject body) {
         try (XContentBuilder builder = XContentFactory.jsonBuilder()) {
             XContentBuilder source = body.toXContent(builder, FOR_INTERNAL_STORAGE_PARAMS);
 
-            IndexRequest indexRequest = new IndexRequest(InferenceIndexConstants.LATEST_INDEX_NAME)
+            return new IndexRequest()
                 .opType(DocWriteRequest.OpType.CREATE)
-                .setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE)
                 .id(docId)
                 .source(source);
-
-            executeAsyncWithOrigin(client, ML_ORIGIN, IndexAction.INSTANCE, indexRequest, indexListener);
         } catch (IOException ex) {
-            // not expected to happen but for the sake of completeness
-            indexListener.onFailure(ex);
+            // This should never happen. If we were able to deserialize the object (from Native or REST) and then fail to serialize it again
+            // that is not the users fault. We did something wrong and should throw.
+            throw new ElasticsearchStatusException("Unexpected serialization exception for [{}]",
```

Comment:
nit: Could use `ExceptionsHelper.serverError`. Not necessary but for future reference.